_studio/mfx_lib/decode/vp9/src/mfx_vp9_dec_decode_hw.cpp

```diff
@@ -85,6 +85,78 @@ bool CheckHardwareSupport(VideoCORE *p_core, mfxVideoParam *p_video_param)
     return true;
 }
 
+class ReferenceFrameStorage
+{
+private:
+    std::vector<UMC_VP9_DECODER::VP9DecoderFrame> m_submittedFrames;
+    mfx_UMC_FrameAllocator *m_frameAllocator;
+
+    void FreeReferenceFrames(UMC::FrameMemID frameId, bool force)
+    {
+        auto find_it = std::find_if(m_submittedFrames.begin(), m_submittedFrames.end(), [frameId](UMC_VP9_DECODER::VP9DecoderFrame frame){ return (frame.currFrame == frameId); });
+        if (find_it != m_submittedFrames.end())
+        {
+            // if force=TRUE need to unlock all reference frames
+            // otherwiseref useless references wil be unlocked in VideoDECODEVP9_HW::UpdateRefFrames()
+            if (force)
+            {
+                for (mfxI32 ref_index = 0; ref_index < NUM_REF_FRAMES; ++ref_index)
+                {
+                    const UMC::FrameMemID mid = find_it->ref_frame_map[ref_index];
+                    if (mid >= 0)
+                    {
+                        m_frameAllocator->DecreaseReference(mid);
+                    }
+                }
+            }
+            m_submittedFrames.erase(find_it);
+        }
+
+    }
+
+public:
+    ReferenceFrameStorage(mfx_UMC_FrameAllocator *frameAllocator):m_frameAllocator(frameAllocator)
+    {
+    }
+
+    void Add(UMC_VP9_DECODER::VP9DecoderFrame currFrame)
+    {
+        m_submittedFrames.push_back(currFrame);
+    }
+
+    void FrameDecoded(UMC::FrameMemID frameId)
+    {
+        auto find_it = std::find_if(m_submittedFrames.begin(), m_submittedFrames.end(), [frameId](UMC_VP9_DECODER::VP9DecoderFrame frame){ return (frame.currFrame == frameId); });
```

Comment:
It would be better to adjust formatting, like it is done at [FreeUnusedReferenceFrames] method.